// decoders/lib/basics.js
'use strict'

exports.__esModule = true
exports.always = always
exports.constant = constant
exports.unknown = exports.undefined_ = exports.optional = exports.nullable = exports.null_ = exports.mixed = exports.maybe = exports.hardcoded = void 0

var _Decoder = require('../Decoder')

var _unions = require('./unions')

var null_ = (0, _Decoder.define)(function (blob, ok, err) {
  return blob === null ? ok(blob) : err('Must be null')
})

exports.null_ = null_
var undefined_ = (0, _Decoder.define)(function (blob, ok, err) {
  return blob === undefined ? ok(blob) : err('Must be undefined')
})
exports.undefined_ = undefined_
var undefined_or_null = (0, _Decoder.define)(function (blob, ok, err) {
  return blob === undefined || blob === null ? ok(blob) : err('Must be undefined or null')
})

function _maybeish(emptyCase) {
  function _inner(decoder) {
    var rv = (0, _unions.either)(emptyCase, decoder)

    if (arguments.length >= 2) {
      var _defaultValue = arguments[1]

      var _defaultValue2 = typeof _defaultValue === 'function' ? _defaultValue() : _defaultValue

      return rv.transform(function (value) {
        return value != null ? value : _defaultValue2
      })
    } else {
      return rv
    }
  }

  return _inner
}

var nullable = _maybeish(null_)

exports.nullable = nullable

var optional = _maybeish(undefined_)

exports.optional = optional

var maybe = _maybeish(undefined_or_null)

exports.maybe = maybe

function constant(value) {
  return (0, _Decoder.define)(function (blob, ok, err) {
    return blob === value ? ok(value) : err('Must be constant ' + String(value))
  })
}

function always(value) {
  return (0, _Decoder.define)(
    typeof value === 'function'
      ? function (
          blob,

          ok,
          _
        ) {
          return ok(value())
        }
      : function (
          blob,

          ok,
          _
        ) {
          return ok(value)
        }
  )
}

var hardcoded = always

exports.hardcoded = hardcoded
var unknown = (0, _Decoder.define)(function (blob, ok, _) {
  return ok(blob)
})

exports.unknown = unknown
var mixed = unknown
exports.mixed = mixed
