// decoders/lib/strings.js
'use strict'

exports.__esModule = true
exports.nonEmptyString = exports.httpsUrl = exports.email = void 0
exports.regex = regex
exports.uuidv4 = exports.uuidv1 = exports.uuid = exports.url = exports.string = void 0

var _Decoder = require('../Decoder')

var _unions = require('./unions')

var _utilities = require('./utilities')

var url_re = /^([A-Za-z]{3,9}(?:[+][A-Za-z]{3,9})?):\/\/(?:([-;:&=+$,\w]+)@)?(?:([A-Za-z0-9.-]+)(?::([0-9]{2,5}))?)(\/(?:[-+~%/.,\w]*)?(?:\?[-+=&;%@.,\w]*)?(?:#[.,!/\w]*)?)?$/

var string = (0, _Decoder.define)(function (blob, ok, err) {
  return typeof blob === 'string' ? ok(blob) : err('Must be string')
})

exports.string = string
var nonEmptyString = regex(/\S/, 'Must be non-empty string')

exports.nonEmptyString = nonEmptyString

function regex(regex, msg) {
  return string.refine(function (s) {
    return regex.test(s)
  }, msg)
}

var email = regex(/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/, 'Must be email')

exports.email = email
var url = (0, _unions.either)(
  regex(url_re, 'Must be URL').transform(function (value) {
    return new URL(value)
  }),
  (0, _utilities.instanceOf)(URL)
)

exports.url = url
var httpsUrl = url.refine(function (value) {
  return value.protocol === 'https:'
}, 'Must be an HTTPS URL')

exports.httpsUrl = httpsUrl
var uuid = regex(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i, 'Must be uuid')

exports.uuid = uuid
var uuidv1 = uuid.refine(function (value) {
  return value[14] === '1'
}, 'Must be uuidv1')

exports.uuidv1 = uuidv1
var uuidv4 = uuid.refine(function (value) {
  return value[14] === '4'
}, 'Must be uuidv4')
exports.uuidv4 = uuidv4
